### PR TITLE
Async package management

### PR DESCRIPTION
**Pull request by dajmaki**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala [hunk 1]
```diff
@@ -118,14 +118,44 @@ object Update {
...
+    */
+
+  final case class PublicPackageUploadRejected(
+      submissionId: SubmissionId,
+      recordTime: Timestamp,
+      rejectionReason: String)
       extends Update {
     override def description: String =
-      s"""Public package uploaded: ${archive.getHash}"""
+      s"Public package upload '$submissionId' rejected: $rejectionReason"
```
Comment (mziolekda) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala]: Please rephrase to:
s"Public package upload rejected, correlationId=$submissionId reason='$rejectionReason'"

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala [hunk 2]
```diff
@@ -60,74 +70,84 @@ class ApiPackageManagementService(
...
-            case r @ UploadPackagesResult.InvalidPackage(_) =>
-              Future.failed(ErrorFactories.invalidArgument(r.description))
-            case r @ UploadPackagesResult.ParticipantNotAuthorized =>
-              Future.failed(ErrorFactories.permissionDenied(r.description))
-            case r @ UploadPackagesResult.NotSupported =>
-              Future.failed(ErrorFactories.unimplemented(r.description))
-          }(DE)
-          .flatMap(pollUntilPersisted(res._2, _))(DE)
+      transactionsService
+        .currentLedgerEnd()
```
Comment (mziolekda) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala]: It seems possible, that a race condition can be won by an upload packages request traveling through the ledger and back through the transaction service. In that case "ledgerEndBeforeRequest" is in fact "ledgerEndAfterRequest"

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryPackageStore.scala [hunk 3]
```diff
@@ -70,6 +73,9 @@ case class InMemoryPackageStore(
     } yield packages
   }
 
+  // FIXME(JM): Move packageEntries into its own interface?
+  def packageEntries(beginOffset: LedgerOffset.Absolute): Source[PackageEntry, NotUsed] = ???
```
Comment (mziolekda) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryPackageStore.scala]: Forgetmenot

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/package_management_service.proto [hunk 4]
```diff
@@ -86,8 +86,16 @@ message UploadDarFileRequest {
   // ``daml_lf.proto``.
   // Required
   bytes dar_file = 1;
+
+  // The deadline for accepting the request after which the ledger rejects it.
+  // Optional, defaults to maximum time to live in ledger's active time model.
+  google.protobuf.Timestamp maximum_record_time = 2;
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/admin/package_management_service.proto]: Agreed. Not in this PR though.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala [hunk 5]
```diff
@@ -118,14 +118,44 @@ object Update {
...
+    *   Description of the upload, if provided by the submitter.
+    * @param recordTime:
+    *   The ledger-provided timestamp at which the package upload was committed.
+    * @param submissionId:
+    *   The submission id of the upload. Unset if this participant was not the submitter.
+    */
+  final case class PublicPackageUpload(
+      archives: List[DamlLf.Archive],
       sourceDescription: Option[String],
-      participantId: ParticipantId,
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala]: Exactly, we don't want to emit these updates if they're for another participant. No need to have it in the Update event if it's never sent. It IMO isn't the indexer's job to filter.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala [hunk 6]
```diff
@@ -118,14 +118,44 @@ object Update {
...
+    * @param recordTime:
+    *   The ledger-provided timestamp at which the package upload was committed.
+    * @param rejectionReason:
+    *   Reason why the upload was rejected.
+    */
+
+  final case class PublicPackageUploadRejected(
+      submissionId: SubmissionId,
+      recordTime: Timestamp,
+      rejectionReason: String)
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala]: See above. I don't agree we need participantIds here.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala (hunk 2 above)
Comment (dajmaki) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala]: Oops, my mistake. Definitely need to query it first.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryPackageStore.scala (hunk 3 above)
Comment (dajmaki) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryPackageStore.scala]: Fixed by dropping this and not extending IndexPackageService (which made no sense anyway).

File: ledger/api-server-damlonx/reference-v2/src/main/scala/com/daml/ledger/api/server/damlonx/reference/v2/ReferenceServer.scala [hunk 7]
```diff
@@ -50,10 +52,11 @@ object ReferenceServer extends App {
   val authService = AuthServiceWildcard
 
   config.archiveFiles.foreach { file =>
+    val submissionId = SubmissionId.assertFromString(UUID.randomUUID().toString)
     for {
       dar <- DarReader { case (_, x) => Try(Archive.parseFrom(x)) }
         .readArchiveFromFile(file)
-    } yield ledger.uploadPackages(dar.all, None)
+    } yield ledger.uploadPackages(submissionId, dar.all, None)
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply ledger/api-server-damlonx/reference-v2/src/main/scala/com/daml/ledger/api/server/damlonx/reference/v2/ReferenceServer.scala]: Not sure I understand what you mean by this?

**Pull request merged**